src/tools/rbd/action/Trash.cc

```diff
@@ -300,6 +317,113 @@ int execute_list(const po::variables_map &vm) {
   return 0;
 }
 
+void get_purge_arguments(po::options_description *positional,
+                            po::options_description *options) {
+  at::add_pool_options(positional, options);
+  at::add_no_progress_option(options);
+  
+  options->add_options()
+      (at::OLDER_THAN.c_str(), po::value<std::string>()->value_name("date"), 
+       "purges images that expired before the given date");
+  options->add_options()
+      (at::THRESHOLD.c_str(), po::value<double>(), "purges images until the current pool data usage is reduced to X%, value range: 0.0-1.0");
+}
+
+
+int execute_purge (const po::variables_map &vm) {
+  size_t arg_index = 0;
+  std::string pool_name = utils::get_pool_name(vm, &arg_index);
+
+  librados::Rados rados;
+  librados::IoCtx io_ctx;
+  int r = utils::init(pool_name, &rados, &io_ctx);
+  if (r < 0) {
+    return r;
+  }
+
+  librbd::RBD rbd;
+  
+  std::vector<trash_info_t> trash_entries;
+  r = rbd.trash_list(io_ctx, trash_entries);
+  if (r < 0) { return r; }
+
+
+  std::vector<const char *> to_be_removed;
+
+  if (vm.find(at::THRESHOLD) != vm.end()) {
+    float threshold = vm[at::THRESHOLD].as<float>();
+    if (threshold < 0 || threshold > 1) {
+        std::cout << "rbd: argument 'threshold' is out of valid range" << std::endl;
+        return -EINVAL;
+    }
+    std::sort(trash_entries.begin(), trash_entries.end(), 
+      [](trash_info_t a, trash_info_t b) { 
+        return a.deferment_end_time < b.deferment_end_time;
+      }
+    );
+
+    std::list<std::string> vec = { pool_name };
+    std::map<std::string,librados::pool_stat_t> stats;
+    rados.get_pool_stats(vec, stats);
+
+    uint64_t bytes_to_free = 0;
+    auto bytes_threshold = (uint64_t)(stats[pool_name].num_bytes * (1 - threshold));
+    
+    librbd::Image curr_img;
+    for (const auto& entry : trash_entries) {
+      r = utils::open_image_by_id(io_ctx, entry.id, true, &curr_img);
+      if(r < 0) continue;
+      
+      uint64_t img_size; curr_img.size(&img_size);
+      r = curr_img.diff_iterate2(nullptr, 0, img_size, false, false, 
```

Comment:
If you change that second ```false``` to ```true``` you will enable fast-diff space calculation which will be much faster at the expense of rounding to 4MB object sizes.